docs/architecture/adr-038-add-support-for-wasm-light-clients.md

```diff
@@ -0,0 +1,86 @@
+
+To remedy these shortcomings, we are proposing a WASM VM to host light client bytecode, which allows easier upgrading of 
+existing light clients as well as adding support for new light clients without requiring a code release and corresponding 
+hard-fork event.
+
+## Context
+Currently in the SDK, light clients are defined as part of the codebase and are implemented as submodules under 
+`/x/ibc/light-clients/`. To add support for new light client or update an existing light client in the event of security 
+issue or consensus update, we need to modify the codebase and integrate it in *numerous* places. In addition, individual 
+blockchains built upon Cosmos SDK need to update all their validator nodes to latest version to add support for this 
+light client. This entire process is tedious and time consuming. In the event a blockchain wants to add support for niche 
+light client, it need to fork Cosmos SDK and modify codebase to add support for it. This creates overhead of maintaining 
+that fork against mainstream Cosmos SDK release.
+
+We are proposing simplifying this workflow by integrating a WASM light client module which makes adding support for 
+new light client a simple transaction. The light client bytecode, written in Wasm-compilable Rust runs inside a Wasmer 
+VM. The Wasm light client submodule exposes a proxy light client interface that routes incoming messages to the 
+appropriate handler function, inside the Wasm VM for execution.
+
+## Decision
+
+We decided to use WASM light client module as a generic light client which will interface with the actual light client 
+uploaded as WASM bytecode. This will require changing client selection method to allow any client if the client type 
+has prefix of `wasm/`.
+
+```go
+// IsAllowedClient checks if the given client type is registered on the allowlist.
+func (p Params) IsAllowedClient(clientType string) bool {
+	for _, allowedClient := range p.AllowedClients {
+		if allowedClient == clientType {
+			return true
+		}
+	}
+	return false || isWASMClient(clientType)
+}
+```
+
+Inside Wasm light client `ClientState`, appropriate Wasm bytecode will be executed depending upon `ClientType`.
+
+```go
+func (cs ClientState) Validate() error {
+    wasmRegistry = getWASMRegistry()
+	clientType := cs.ClientType()
+    codeHandle := wasmRegistry.getCodeHandle(clientType)
+    return codeHandle.validate(cs)
+}
+```
+
+To upload new light client, user need to create a transaction with Wasm byte code and unique client id, which will be 
+processed by IBC Wasm module.
+
+```go
+func (k Keeper) UploadLightClient (WasmCode: []byte, id: String, description: String) {
+    wasmRegistry = getWASMRegistry()
+    assert(!wasmRegistry.exists(id))
+    assert(wasmRegistry.validateAndStoreCode(id, description, wasmCode, false))
+}
+```
+
+As name implies, Wasm registry is a registry which stores set of Wasm client code indexed by its unique id and allows 
+client code to retrieve latest code uploaded.
```

Comment:
I think we should be also able to retire / disable the old light client.

Reply:
I agree with this

Reply:
Perhaps a client type can be deleted if no clients using that type exist anymore.

However, at the moment, IBC light clients themselves cannot be deleted so this is not so useful.

@robert-zaremba Did you have a particular process in mind?

Reply:
We cannot disable/retire previous version of light client, if there are existing clients using that. New clients will only be created with latest version of Light client. 

In future, we can add `upgrade` process, but it is out of scope right now.

Reply:
I still think it should be a part of the "standard". Maybe we can mark it as an optional.  The motivation for standard is: we will have many light clients, and light client users will need to understand some basic API about them. The motivation for disabling a light-client is that if there is a security bug, then we are better to switch it off, then corrupting the chain.

The process would be:

* there is a gov smart-contract (could PoA or a democracy...) which has an access to set light client flags:
  * `deprecated`
  * `disabled` (optional).
* the gov contract firstly sets deprecated flag, then it can disable (optionally) the light client. If disabling, we can even remove it from the state.